### ISSUE TITLE
Tpetra::CrsGraph::getLocalDiagOffsets: Kokkos-ize

### ISSUE THREAD

**Issue by mhoemmen** [trilinos/Trilinos#212]
This blocks #41, but is only a serious blocker if the matrix's graph structure changes frequently.
This depends on #205.

It's OK for `Tpetra::CrsGraph::getLocalDiagOffsets` only to have a full Kokkos parallelization (e.g., on CUDA device) if the graph is fillComplete.  If not, either the existing sequential code or a host execution space parallelization suffice.  Thus, the code in CrsGraph should branch between these two cases.

Ifpack2 should only exercise the fillComplete case, otherwise it's a performance bug for Ifpack2 and possibly a correctness bug for the application.  (Giving a non-fillComplete matrix to Ifpack2 is wrong, because one should always fillComplete the matrix before handing it off to a solver.  If the matrix has never been fillComplete before, calling fillComplete could change the offsets of the diagonal entries, which is what getLocalDiagOffsets computes.)

Take care to make the whole Kokkos::StaticCrsGraph memory-unmanaged before entering the Kokkos kernel.  Otherwise, it will be slow with non-Cuda execution spaces, due to memory management overhead.

**Comment by mhoemmen** [comment]
Attached here is a patch that gets you part of the way to a fix for this issue.  Please apply after the patches for #205, as it depends on those (in particular, it uses findRelOffset).

[0001-TMP-Tpetra-CrsGraph-getLocalDiagOffsets-Partway-to-2.txt](https://github.com/trilinos/Trilinos/files/170344/0001-TMP-Tpetra-CrsGraph-getLocalDiagOffsets-Partway-to-2.txt)

By "part of the way," I mean that every function called in the fillComplete branch is a Kokkos device function, and every Kokkos::View used there is a View of device memory.  Thus, the only thing left to do is turn it into a functor and drop it into a parallel_for.  The elaborate debug-build error checking could be left out for now.  (For extra credit, in a debug build, do some error checking and use a parallel_reduce to find any errors.  Omit the std::vector::push_back tracking.)

**Comment by mhoemmen** [comment]
@amklinv -- I assigned you provisionally, but please note that this is lower priority than the two-argument version of getLocalDiagCopy.  I just want to upload my patches today while I still can.

**Comment by mhoemmen** [comment]
The patches in this comment supersede my patch above.  The first four patches come from my comments to #205.  The last patch implements a Kokkos functor that purports to fix this issue.  It builds and passes Kokkos::Serial tests.  I have _not_ yet tested it with CUDA.  Really, I just want to upload the patches so that they don't get lost over the next few days.

[0001-Kokkos-StaticCrsGraph-Add-findRelOffset-see-Trilinos.txt](https://github.com/trilinos/Trilinos/files/170400/0001-Kokkos-StaticCrsGraph-Add-findRelOffset-see-Trilinos.txt)
[0002-Tpetra-CrsGraph-Add-test-for-new-findRelOffset-see-2.txt](https://github.com/trilinos/Trilinos/files/170401/0002-Tpetra-CrsGraph-Add-test-for-new-findRelOffset-see-2.txt)
[0003-TMP-Tpetra-CrsGraph-findLocalIndex-Use-findRelOffset.txt](https://github.com/trilinos/Trilinos/files/170404/0003-TMP-Tpetra-CrsGraph-findLocalIndex-Use-findRelOffset.txt)
[0004-Kokkos-StaticCrsGraph-Add-binary-search-to-findRelOf.txt](https://github.com/trilinos/Trilinos/files/170402/0004-Kokkos-StaticCrsGraph-Add-binary-search-to-findRelOf.txt)
[0005-Tpetra-CrsGraph-getLocalDiagOffsets-Fix-212.txt](https://github.com/trilinos/Trilinos/files/170403/0005-Tpetra-CrsGraph-getLocalDiagOffsets-Fix-212.txt)

**Comment by mhoemmen** [comment]
OK, final set of patches for now.  These three patches supersede all of the above patches.  I think only the first two are necessary, but the third won't hurt.

[0001-Tpetra-Add-findRelOffset-to-fix-205.txt](https://github.com/trilinos/Trilinos/files/170441/0001-Tpetra-Add-findRelOffset-to-fix-205.txt)
[0002-Tpetra-CrsGraph-getLocalDiagOffsets-Fix-212.txt](https://github.com/trilinos/Trilinos/files/170443/0002-Tpetra-CrsGraph-getLocalDiagOffsets-Fix-212.txt)
[0003-TMP-Tpetra-CrsGraph-findLocalIndex-Use-findRelOffset.txt](https://github.com/trilinos/Trilinos/files/170442/0003-TMP-Tpetra-CrsGraph-findLocalIndex-Use-findRelOffset.txt)

**Comment by mhoemmen** [comment]
My recent push includes all of the above patches, _except_ for 0003-TMP-Tpetra-CrsGraph-findLocalIndex-Use-findRelOffset.txt.
